lib/tree.rb

```diff
@@ -16,40 +16,127 @@ def initialize
     @root = nil
   end
 
+  def add_helper(current_node, key, value)
+    # base case is when current node is nil
+    return TreeNode.new(key, value) if current_node.nil?
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end
+
   # Time Complexity: 
   # Space Complexity: 
   def add(key, value)
-    raise NotImplementedError
+    # recursive solution with helper method
+    @root = add_helper(@root, key, value)
+
+    # the while-looping way
+    # if @root.nil?
+    #   @root = TreeNode.new(key, value)
+    # else
+    #   current = @root
+    #   while true
+    #     if key < current.key
+    #       if !current.left.nil?
+    #         current = current.left
+    #       else
+    #         current.left = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     else
+    #       if !current.right.nil?
+    #         current = current.right
+    #       else
+    #         current.right = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     end
+    #   end
+    # end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+
+    if key < current_node.key
+      find_helper(current_node.left, key)
+    else
+      find_helper(current_node.right, key)
+    end
+  end
+
+  # Time Complexity: O(logn) where n is the number of nodes
+  # Space Complexity: O(logn) where n is the number of nodes
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    return list
+  end
   # Time Complexity: 
   # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    # left -> current node -> right
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    # root -> left -> right
+    return list if current_node.nil?
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(logn) where n is the number of nodes
+  # Space Complexity: O(logn) where n is the number of nodes
   def preorder
```

Comment:
Actually since you hit every node in the tree, the time complexity is O(n).  The maximum recursive calls (assuming balanced tree) is O(log n), for space complexity.

Otherwise this works.